### DISCUSSION TITLE
RTLMP rotation of macros?

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#4524] [General]
Does RTLMP rotate macros?

I have purposely created macros with all pins (except clock) on the left side so that there is an asymmetry to be exploited.

Looking at the register file, highlighted in the screenshot below, it looks like rotating it 90 degrees counterclockwise would have resulted in much reduced routing congestion compared to the MY (mirrored about the Y axis) orientation.

I have seen R180, but I don't recall R90... Does RTLMP try R90?

It is incredibly nice to get an estimated routing congestion at the floorplan stage. Thanks @ApeachM!

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/fffc3384-833f-4141-ab8d-b5dbc4be1d47)

This was using https://github.com/The-OpenROAD-Project/megaboom/commits/pins-on-left/

Standalone reproduction case:

1. Compile #4519
1. untar https://drive.google.com/file/d/1bzlI2F004DV0dn4uvH0ajKE7l7bl9yW2/view?usp=sharing
2. run `run-me-ChipTop-asap7-base.sh`

**Comment by maliberty** [comment]
No we don't do 90 degree turns.  In most advanced processes that is illegal as there is a required poly direction.

**Reply by oharboe** [reply]
Makes sense.

**Reply by oharboe** [reply]
@maliberty I modified the macros so that some have pins on top and some on the left edge. This seems to fix the macro related global routing congestion. What I think is going on here is simply that the placement density is simply too high for this design.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/c4de96f9-c9f0-4ad1-a76e-9db83d54d1ea)
